docs/developer/go-style-guide.md

```diff
@@ -0,0 +1,458 @@
+# Go Style Guide
+
+Before diving into project-specific guidelines, please familiarize yourself with the following vetted best practices:
+
+- [Effective Go](https://go.dev/doc/effective_go)
+- [Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
+- [100 Go Mistakes](https://github.com/teivah/100-go-mistakes)
+
+Once you have a good grasp of these general best practices, you can then explore the project-specific guidelines below.
+These guidelines will often build upon the foundation set by the general best practices and provide additional
+recommendations tailored to the project's specific requirements and coding style.
+
+## Table of Contents
+
+1. [General Guidelines](#general-guidelines)
+2. [Error Handling](#error-handling)
+3. [Logging](#logging)
+4. [Concurrency](#concurrency)
+5. [Recommended/Situational Guidelines](#recommended--situational)
+
+## General Guidelines
+
+### Use the empty struct `struct{}` for sentinel values
+
+Empty structs as sentinels unambiguously signal an explicit lack of information. For example, use empty struct when for
+sets and for signaling via channels that don't require a message.
+
+DO:
+
+```go
+set := make(map[string]struct{}) // empty struct is nearly value-less
+
+signaller := make(chan struct{}, 0)
+signaller <- struct{}{} // no information but signal on delivery
+```
+
+DO NOT:
+
+```go
+set := make(map[string]bool) // is true/false meaningful? Is this a set?
+
+signaller := make(chan bool, 0)
+signaller <- false // is this a signal? is this an error?
+```
+
+### Consistent Line Breaks
+
+When breaking up a long function definition, call, or struct initialization, choose to break after each parameter,
+argument, or field.
+
+DO:
+
+```go
+func longFunctionDefinition(
+    paramX int,
+    paramY string,
+    paramZ bool,
+) (string, error){}
+
+// and 
+
+s := myStruct{
+    field1: 1,
+    field2: 2,
+    field3: 3,
+}
+```
+
+DO NOT:
+
+```go
+func longFunctionDefinition(paramX int, paramY string,
+    paramZ bool,
+) (string, error){}
+
+// or 
+
+func longFunctionDefinition(
+    paramX int, paramY string,
+    paramZ bool,
+) (string, error){}
+
+// or 
+s := myStruct{field1: 1, field2: 2,
+    field3: 3}
+
+```
+
+When constructing structs pass members during initialization.
+
+Example:
+
+```go
+cfg := foo.Config{
+    Site: "example.com",
+    Out: os.Stdout,
+    Dest: c.KeyPair{
+        Key: "style",
+        Value: "well formatted",
+    },
+}
+```
+
+### Do not copy sync entities
+
+`sync.Mutex` and `sync.Cond` MUST NOT be copied. By extension, structures holding an instance MUST NOT be copied, and
+structures which embed instances MUST NOT be copied.
+
+DO NOT embed sync entities. Pointers to `Mutex` and `Cond` are required for storage.
+
+### Construct slices with known capacity
+
+Whenever possible bounded slices should be constructed with a length of zero size, but known capacity.
```

Comment:
Our `makezero` linter *should* guard against this.

Reply:
I don't think I've seen the linter complain about this before. There are a couple of places in our code base where it's impossible to know the capacity of the slice and the linter doesn't flag them..

Reply:
Yeah, it should only flag if a capacity can be known. But if we're not confident in it, I'm fine leaving this in the doc.